nn_surrogate_benchmark/surrogate_sobol.py

```diff
@@ -0,0 +1,209 @@
+import torch
+import torch.nn as nn
+import pytorch_lightning as pl
+from torch.optim import AdamW, Adam
+from typing import Literal
+import pandas as pd
+import numpy as np
+from torch.utils.data import TensorDataset, DataLoader
+from sklearn.preprocessing import (
+	StandardScaler,
+	MinMaxScaler,
+	RobustScaler,
+	MaxAbsScaler,
+)
+
+class Sobolev(pl.LightningModule):
+	def __init__(
+		self,
+		input_dim: int,
+		hidden_dims: list[int] = [512],
+		activation: Literal["tanh", "relu", "gelu"] = "tanh",
+		lam1: float = 1.0,  # Weighting factor for gradient loss
+		lr = 3e-4,
+	) -> None:
+		super().__init__()
+		self.save_hyperparameters()
+
+		layers = []
+		for hidden_dim in hidden_dims:
+			layers.append(nn.Linear(input_dim, hidden_dim))
+			if activation == "tanh":
+				layers.append(nn.Tanh())
+			elif activation == "relu":
+				layers.append(nn.ReLU(inplace=False))  # Ensure inplace=False
+			else:
+				layers.append(nn.GELU())
+			input_dim = hidden_dim
+		layers.append(nn.Linear(input_dim, 1))
+		self.net = nn.Sequential(*layers)
+		self.criterion = nn.MSELoss()
+		self.net.apply(self._init_weights)
+
+
+	def _init_weights(self, module):
+		if isinstance(module, nn.Linear):
+			if self.hparams.activation in ["relu", "gelu"]:
+				nn.init.kaiming_normal_(module.weight, nonlinearity="relu")
+			else:
+				nn.init.xavier_normal_(
+					module.weight, gain=nn.init.calculate_gain(self.hparams.activation)
+				)
+
+			if module.bias is not None:
+				nn.init.zeros_(module.bias)
+
+	def forward(self, x):
+		return self.net(x)
+
+	def training_step(self, batch, batch_idx):
+		x, y, dy_dx = batch
+		x = x.detach().clone().requires_grad_(True)
+
+		# Forward pass
+		preds = self(x)
+
+		# Compute gradients using .backward()
+		grads = torch.autograd.grad(
+			outputs=preds,
+			inputs=x,
+			grad_outputs=torch.ones_like(preds),
+			create_graph=True,
+			retain_graph=True,
+		)[0]
+
+		# Compute losses
+		loss_value = self.criterion(preds, y)
+		loss_grad = self.criterion(grads, dy_dx)
+		total_loss = loss_value + self.hparams.lam1 * loss_grad
+
+		# Log losses
+		self.log("train_loss", total_loss, prog_bar=True, on_epoch=True)
+		self.log("train_value_loss", loss_value, on_epoch=True)
+		self.log("train_grad_loss", loss_grad, on_epoch=True)
+
+		return total_loss
+
+	def validation_step(self, batch, batch_idx):
+		x, y = batch
+		preds = self(x)
+		loss = self.criterion(preds, y)
+		self.log("val_loss", loss, prog_bar=True, on_epoch=True)
+		return loss
+
+	def test_step(self, batch, batch_idx):
+		x, y = batch
+		preds = self(x)
+		loss = self.criterion(preds, y)
+		self.log("test_loss", loss, on_epoch=True, prog_bar=True)
+		return loss
+
+	def configure_optimizers(self):
+		optimizer = Adam(self.parameters(), lr=self.hparams.lr, weight_decay=1e-5)
+		scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=300, gamma=0.5)
```

Comment:
for MLP we have:
```
    def configure_optimizers(self):
        optimizer = AdamW(self.parameters(), lr=self.hparams.lr)
        scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(
            optimizer, mode="min", factor=0.1, patience=5, verbose=True
        )

        return {
            "optimizer": optimizer,
            "lr_scheduler": {"scheduler": scheduler, "monitor": "val_loss"},
        }
```
why do use StepLR here? Does it work better?

Reply:
Yes, I arrived at StepLR through trial and error